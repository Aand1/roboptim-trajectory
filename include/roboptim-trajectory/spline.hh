// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Spline declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_SPLINE_HH
# define ROBOPTIM_TRAJECTORY_SPLINE_HH
# include <roboptim-trajectory/trajectory.hh>

# include <roboptim-trajectory/fwd.hh>

class bspline;

namespace roboptim
{
  class Spline : public Trajectory<4>
  {
  public:
    Spline (size_type, const vector_t&, int, int, int) throw ();
    virtual ~Spline () throw ();

    virtual vector_t operator () (double) const throw ();

    virtual vector_t derivative (double x, size_type order) const throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

  private:
    bspline* spline_;
  };
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
