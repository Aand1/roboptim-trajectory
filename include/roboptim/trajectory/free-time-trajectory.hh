// Copyright (C) 2009 by Florent Lamiraux, Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HH
# define ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HH
# include <roboptim/trajectory/trajectory.hh>

namespace roboptim
{
  template <unsigned DerivabilityOrder>
  class FreeTimeTrajectory : public Trajectory<DerivabilityOrder>
  {
  public:
    /// Constructor with fixed definition interval trajectory
    ///
    /// \param traj trajectory defining this one by reparameterization
    /// \param s time scale
    FreeTimeTrajectory (const Trajectory<DerivabilityOrder>& traj, double s) throw ();
    virtual ~FreeTimeTrajectory () throw ();

    virtual vector_t operator () (double) const throw ();

    virtual vector_t derivative (double x, size_type order) const throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

    /// Get index of time scaling parameter in vector
    double timeScale () const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  private:
    double scaleTime (double) const throw ();

    /// Fixed time trajectory
    const Trajectory<DerivabilityOrder>& trajectory_;
    /// Store time scaling parameter \f$\textbf{p}_{m+1}\f$.
    double timeScale_;
  };
} // end of namespace roboptim.

# include <roboptim/trajectory/free-time-trajectory.hxx>
#endif //! ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HH
