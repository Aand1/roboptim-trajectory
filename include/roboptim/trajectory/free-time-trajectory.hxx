// Copyright (C) 2009 by Florent Lamiraux, Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HXX
# define ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HXX
# include <cmath>

# include <boost/numeric/ublas/matrix.hpp>
# include <boost/numeric/ublas/matrix_proxy.hpp>

namespace roboptim
{
  namespace detail
  {
    inline Function::value_type
    scaleTime (Function::value_type unscaled,
	       Function::value_type min,
	       Function::value_type scale)
    {
      return min + (unscaled - min) / scale;
    }

    template <unsigned dorder>
    Function::interval_t
    scaleInterval (const Trajectory<dorder>& traj,
		   typename Function::value_type scale)
    {
      Function::value_type min =
	Function::getLowerBound (traj.timeRange ());
      Function::value_type max =
	scaleTime (Function::getUpperBound (traj.timeRange ()), min, scale);
      return Function::makeInterval (min, max);
    }
  } // end of namespace detail

  template <unsigned dorder>
  FreeTimeTrajectory<dorder>::FreeTimeTrajectory
  (const Trajectory<dorder>& traj, value_type s) throw ()
    : Trajectory<dorder> (detail::scaleInterval (traj, s), traj.outputSize (),
			  addScaleToParameters (traj.parameters (), s)),
      trajectory_ (traj.clone ())
  {
    assert (s != 0. && !std::isinf (s) && !std::isnan (s));
  }

  template <unsigned dorder>
  FreeTimeTrajectory<dorder>::FreeTimeTrajectory
  (const FreeTimeTrajectory<dorder>& traj)
    throw ()
    : Trajectory<dorder> (traj.timeRange (), traj.outputSize (),
			  traj.parameters ()),
      trajectory_ (traj.trajectory_->clone ())
  {
  }

  template <unsigned dorder>
  FreeTimeTrajectory<dorder>::~FreeTimeTrajectory () throw ()
  {
    delete trajectory_;
  }

  template <unsigned dorder>
  void
  FreeTimeTrajectory<dorder>::impl_compute
  (typename FreeTimeTrajectory<dorder>::result_t& res , double t) const throw ()
  {
    (*trajectory_) (res, this->scaleTime (t));
  }

  template <unsigned dorder>
  void
  FreeTimeTrajectory<dorder>::impl_derivative
  (typename FreeTimeTrajectory<dorder>::gradient_t& derivative,
   double t,
   typename FreeTimeTrajectory<dorder>::size_type order) const throw ()
  {
    assert (order >= 0);
    double scaled = this->scaleTime (t);
    trajectory_->derivative (derivative, scaled, order);
    derivative *= std::pow (this->timeScale (), 0. + order);
  }

  template <unsigned dorder>
  typename FreeTimeTrajectory<dorder>::jacobian_t
  FreeTimeTrajectory<dorder>::variationConfigWrtParam (double t) const throw ()
  {
    using namespace boost::numeric::ublas;
    value_type scaled = this->scaleTime (t);

    double tMin = this->getLowerBound (this->trajectory_->timeRange ());

    jacobian_t result (this->outputSize (),
		       this->parameters ().size ());
    result.clear ();

    // Compute variation w.r.t time scale (p_0)
    column (result, 0) = (t - tMin) * trajectory_->derivative (scaled, 1);

    // Fill 1..(n-1) lines with original jacobian.
    project (result, range (0, result.size1 ()), range (1, result.size2 ()))
      = trajectory_->variationConfigWrtParam (scaled);

    return result;
  }

  template <unsigned dorder>
  typename FreeTimeTrajectory<dorder>::jacobian_t
  FreeTimeTrajectory<dorder>::variationDerivWrtParam (double t, size_type order)
    const throw ()
  {
    if (order == 0)
      return this->variationConfigWrtParam (t);

    using namespace boost::numeric::ublas;
    value_type scaled = this->scaleTime (t);

    double tMin = this->getLowerBound (this->trajectory_->timeRange ());

    jacobian_t result (this->outputSize (),
		       this->parameters ().size ());
    result.clear ();

    // Compute variation w.r.t time scale (p_0)
    column (result, 0) = trajectory_->derivative (scaled, 1) * order;
    column (result, 0) += trajectory_->derivative (scaled, 1)
      * this->timeScale () * (t - tMin);
    column (result, 0) *= std::pow (this->timeScale (), order - 1.);

    // Fill 1..(n-1) lines with original jacobian.
    project (result, range (0, result.size1 ()), range (1, result.size2 ()))
      = trajectory_->variationDerivWrtParam (scaled, order)
      * std::pow (this->timeScale (), 0. + order);

    return result;
  }

  template <unsigned dorder>
  typename FreeTimeTrajectory<dorder>::value_type
  FreeTimeTrajectory<dorder>::singularPointAtRank (size_type rank) const
  {
    double tMin = this->getLowerBound (this->timeRange ());
    return tMin + (trajectory_->singularPointAtRank (rank) - tMin)
      * this->timeScale ();
  }

  template <unsigned dorder>
  typename FreeTimeTrajectory<dorder>::vector_t
  FreeTimeTrajectory<dorder>::derivBeforeSingularPoint (size_type rank,
							size_type order)
    const
  {
    return trajectory_->derivBeforeSingularPoint (rank, order);
  }

  template <unsigned dorder>
  typename FreeTimeTrajectory<dorder>::vector_t
  FreeTimeTrajectory<dorder>::derivAfterSingularPoint (size_type rank, size_type order)
    const
  {
    return trajectory_->derivAfterSingularPoint (rank, order);
  }

  template <unsigned dorder>
  void
  FreeTimeTrajectory<dorder>::setParameters (const vector_t& p) throw ()
  {
    //FIXME: is this ok?
    vector_t p_ = p;
    if (p_[0] <= 0.)
      p_[0] = 1e-8;

    this->parameters_ = p_;
    this->timeRange_ = detail::scaleInterval (*trajectory_, this->timeScale ());
    this->trajectory_->setParameters (removeScaleFromParameters (p_));
  }

  template <unsigned dorder>
  typename FreeTimeTrajectory<dorder>::value_type
  FreeTimeTrajectory<dorder>::timeScale () const throw ()
  {
    return this->parameters_[0];
  }

  template <unsigned dorder>
  double
  FreeTimeTrajectory<dorder>::scaleTime (double unscaled) const throw ()
  {
    value_type tMin = getLowerBound (this->timeRange ());
    value_type tmin = getLowerBound (this->trajectory_->timeRange ());
    value_type tmax = getUpperBound (this->trajectory_->timeRange ());

    unscaled = detail::fixTime (unscaled, *this);
    assert (this->isValidTime (unscaled));

    value_type res = tmin + (unscaled - tMin) * timeScale ();

    if (res > tmax)
      res = tmax;
    else if (res < tmin)
      res = tmin;
    return res;
  }

  template <unsigned dorder>
  double
  FreeTimeTrajectory<dorder>::unscaleTime (double scaled) const throw ()
  {
    value_type tMin = getLowerBound (this->timeRange ());
    value_type tMax = getUpperBound (this->timeRange ());
    value_type tmin = getLowerBound (this->trajectory_->timeRange ());

    scaled = detail::fixTime (scaled, *this);
    assert (trajectory_->isValidTime (scaled));

    value_type res = tMin + (scaled - tmin) / timeScale ();

    if (res > tMax)
      res = tMax;
    else if (res < tMin)
      res = tMin;
    return res;
  }

  template <unsigned dorder>
  std::ostream&
  FreeTimeTrajectory<dorder>::print (std::ostream& o) const throw ()
  {
    o << "Free time trajectory." << std::endl;
    return o;
  }
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HXX
