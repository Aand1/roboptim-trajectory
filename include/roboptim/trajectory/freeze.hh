// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREEZE_HH
# define ROBOPTIM_TRAJECTORY_FREEZE_HH
# include <roboptim/trajectory/sys.hh>

# include <vector>
# include <utility>

# include <roboptim/core/fwd.hh>
# include <roboptim/core/problem.hh>

namespace roboptim
{
namespace trajectory
{
  /// \brief Add constraints that freeze parameters.
  ///
  /// This constraint expects a vector of pairs (argument id, value) to
  /// build a function that will freeze the given arguments to their
  /// associated value.
  ///
  /// For instance, the vector: [(0, 5.), (3, -12.)] forces the first
  /// parameter to five and the fourth one to minus twelve.
  ///
  /// This adds to the problem one linear constraint per frozen parameter.
  template <typename P>
  class Freeze
  {
  public:
    /// \brief Pair representing an argument index and a value.
    typedef std::pair<Function::size_type,
		      Function::value_type> frozenArgument_t;

    /// \brief Problem type.
    typedef P problem_t;

    /// \brief Vector of pairs (argument index, value).
    ///
    /// This type define what are the frozen arguments and what their
    /// value.
    typedef std::vector<frozenArgument_t> frozenArguments_t;

    /// \brief Create the constraint from a vector of pairs.
    ///
    /// \param problem problem that will be modified.
    Freeze (problem_t& problem);

    virtual ~Freeze ();

    /// \brief Apply modification.
    /// \param fa Vector of pairs containing what to freeze and to what value.
    void operator () (const frozenArguments_t fa);

    /// \brief Apply modification.
    /// \param indices Vector of parameters index that will be frozen.
    /// \param values Vector of parameters values.
    void operator () (const std::vector<Function::size_type>& indices,
		      const Function::vector_t& values);

  private:
    /// \brief Reference to the problem that will be modified.
    problem_t& problem_;
  };

  template <typename P>
  Freeze<P> makeFreeze (P& problem)
  {
    return Freeze<P> (problem);
  }

  /// Example shows freeze use.
  /// \example spline-optimization.cc

} // end of namespace trajectory.
} // end of namespace roboptim.

# include <roboptim/trajectory/freeze.hxx>
#endif //! ROBOPTIM_TRAJECTORY_FREEZE_HH
