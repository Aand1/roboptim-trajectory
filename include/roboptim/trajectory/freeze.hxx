// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREEZE_HXX
# define ROBOPTIM_TRAJECTORY_FREEZE_HXX
# include <roboptim/core/numeric-linear-function.hh>

namespace roboptim
{
  template <typename P>
  Freeze<P>::Freeze (problem_t& problem)
    : problem_ (problem)
  {
  }

  template <typename P>
  Freeze<P>::~Freeze ()
  {
  }

  template <typename P>
  void
  Freeze<P>::operator () (const frozenArguments_t frozenArguments)
  {
    using namespace boost;
    typedef frozenArguments_t::const_iterator citer_t;

    for (citer_t it = frozenArguments.begin ();
	 it != frozenArguments.end (); ++it)
      {
	assert (it->first < problem_.function ().inputSize ());

	Function::interval_t& interval =
	  this->problem_.argumentBounds()[it->first];

	Function::value_type min =
	  std::max (Function::getLowerBound (interval), it->second);
	Function::value_type max =
	  std::min (Function::getUpperBound (interval), it->second);
	this->problem_.argumentBounds()[it->first] = Function::makeInterval (min, max);
      }
  }

  template <typename P>
  void
  Freeze<P>::operator () (const std::vector<Function::size_type>& indices,
			  const Function::vector_t& values)
  {
    frozenArguments_t fa;
    for (unsigned i = 0; i < indices.size (); ++i)
      {
	assert (indices[i] < values.size ());
	fa.push_back (std::make_pair (indices[i], values[indices[i]]));
      }
    (*this) (fa);
  }

} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_FREEZE_HXX
