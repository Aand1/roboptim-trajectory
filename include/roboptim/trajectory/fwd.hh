// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FWD_HH
# define ROBOPTIM_TRAJECTORY_FWD_HH
# include <roboptim/trajectory/sys.hh>

namespace roboptim
{
  template <typename T>
  class FreeTimeTrajectory;

  template <unsigned dorder>
  class Trajectory;

  template <typename T>
  class StateFunction;

  template <typename T>
  class SumCost;

  template <typename T>
  class TrajectoryCost;

  class Spline;
  class SplineLength;
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_FWD_HH
