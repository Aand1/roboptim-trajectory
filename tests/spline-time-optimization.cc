// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <fstream>

#include <boost/assign/list_of.hpp>
#include <boost/mpl/vector.hpp>
#include <boost/numeric/ublas/io.hpp>

#include <roboptim/core/finite-difference-gradient.hh>
#include <roboptim/core/solver-factory.hh>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/visualization/gnuplot-function.hh>

#include <roboptim/trajectory/free-time-trajectory.hh>
#include <roboptim/trajectory/freeze.hh>
#include <roboptim/trajectory/fwd.hh>
#include <roboptim/trajectory/limit-speed.hh>
#include <roboptim/trajectory/spline-length.hh>
#include <roboptim/trajectory/spline.hh>
#include <roboptim/trajectory/trajectory-cost.hh>

#include <roboptim/trajectory/visualization/limit-speed.hh>


#include "common.hh"

using namespace roboptim;
using namespace roboptim::visualization;
using namespace roboptim::visualization::gnuplot;

typedef boost::mpl::vector<DerivableFunction, LinearFunction> constraint_t;
typedef Solver<DerivableFunction, constraint_t> solver_t;
typedef FreeTimeTrajectory<Spline::derivabilityOrder> freeTime_t;


// Problem parameters.
const unsigned nControlPoints = 31;
const double vMax = 200.;

int run_test ()
{
  using namespace boost;
  using namespace boost::assign;

  const double finalPos = 200.;
  Spline::vector_t params (nControlPoints);

  for (unsigned i = 0; i < nControlPoints; ++i)
    params[i] = finalPos / (nControlPoints - 1) * i;

  // Make trajectories.
  Spline::interval_t timeRange = Spline::makeInterval (0., 4.);
  Spline spline (timeRange, 1, params, "before");
  freeTime_t freeTimeTraj (spline, 1.);

  // Define cost.
  Function::matrix_t a (1, freeTimeTraj.parameters ().size ());
  a.clear ();
  a (0, 0) = 1.;
  Function::vector_t b (1);
  b.clear ();
  roboptim::NumericLinearFunction cost (a, b);

  // Create problem.
  solver_t::problem_t problem (cost);
  problem.startingPoint () = freeTimeTraj.parameters ();

  {
    // Be sure that scale is positive.
    Function::matrix_t a (1, problem.function ().inputSize ());
    Function::vector_t b (1);
    a.clear (), b.clear ();
    a (0, 0) = 1.;
    shared_ptr<NumericLinearFunction> speedPositivity
      (new NumericLinearFunction (a, b));
    problem.addConstraint
      (static_pointer_cast<LinearFunction> (speedPositivity),
       Function::makeLowerInterval (1e-3));
  }

  const freeTime_t::vector_t freeTimeParams = freeTimeTraj.parameters ();
  const unsigned freeTimeParamsSize = freeTimeParams.size ();

  typedef Freeze<DerivableFunction, constraint_t, DerivableFunction> freeze_t;
  freeze_t freeze (problem,
		   list_of <freeze_t::frozenArgument_t>
		   (1, freeTimeParams[1])
		   (freeTimeParamsSize - 1,
		    freeTimeParams[freeTimeParamsSize - 1]));
  freeze ();

  Function::interval_t vRange (0., 2 * vMax * vMax);
  LimitSpeed<FreeTimeTrajectory<Spline::derivabilityOrder> >::addToProblem
    (freeTimeTraj, problem, vRange, nControlPoints * 10);

  std::ofstream limitSpeedStream ("limit-speed.gp");
  Gnuplot gnuplot = Gnuplot::make_interactive_gnuplot ();

  gnuplot
    << set ("multiplot layout 1,2 title "
	    "'variation of speed before and after optimization'")
    << set ("grid");
  gnuplot << plot_limitSpeed (freeTimeTraj, vMax);


  SolverFactory<solver_t> factory ("cfsqp", problem);
  solver_t& solver = factory ();

  std::cout << solver << std::endl;

  solver_t::result_t res = solver.minimum ();
  std::cout << res << std::endl;

  FreeTimeTrajectory<Spline::derivabilityOrder> optimizedTrajectory =
    freeTimeTraj;

  switch (solver.minimumType ())
    {
    case GenericSolver::SOLVER_VALUE:
      {
	const Result& result = solver.getMinimum<Result> ();
	optimizedTrajectory.setParameters (result.x);
	break;
      }

    case GenericSolver::SOLVER_VALUE_WARNINGS:
      {
	const ResultWithWarnings& result =
	  solver.getMinimum<ResultWithWarnings> ();
	optimizedTrajectory.setParameters (result.x);
	break;
      }

    case GenericSolver::SOLVER_NO_SOLUTION:
    case GenericSolver::SOLVER_ERROR:
      return 1;
    }

  gnuplot << plot_limitSpeed (optimizedTrajectory, vMax);
  limitSpeedStream << (gnuplot << unset ("multiplot"));
  return 0;
}

GENERATE_TEST ()
