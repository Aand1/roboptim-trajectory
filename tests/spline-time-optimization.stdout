Solver:
  Plugin: ipopt
  Problem:
    Numeric linear function
      A = [7](-1,0,0,0,0,0,0)
      B = [1](0)
    Arguments bounds: (0, inf), (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
    Arguments scales: 1, 1, 1, 1, 1, 1, 1
    Number of constraints: 32
    Constraint 0
        Numeric linear function
          A = [7](0,0.166667,0.666667,0.166667,0,0,0)
          B = [1](-0)
        Bounds: (0, 0)
        Scales: 1
        Initial value: [1](0)
    Constraint 1
        Numeric linear function
          A = [7](0,0,0,0,0.166667,0.666667,0.166667)
          B = [1](-200)
        Bounds: (0, 0)
        Scales: 1
        Initial value: [1](-5.68434e-14) (constraint not satisfied)
    Constraint 2
        speed limit (0.0322581) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](0.246678)
    Constraint 3
        speed limit (0.0645161) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](3.94685)
    Constraint 4
        speed limit (0.0967742) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](19.9809)
    Constraint 5
        speed limit (0.129032) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](63.1496)
    Constraint 6
        speed limit (0.16129) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](154.174)
    Constraint 7
        speed limit (0.193548) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](319.695)
    Constraint 8
        speed limit (0.225806) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](592.274)
    Constraint 9
        speed limit (0.258065) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](1010.39)
    Constraint 10
        speed limit (0.290323) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](1618.46)
    Constraint 11
        speed limit (0.322581) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](2466.78)
    Constraint 12
        speed limit (0.354839) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](3532.46)
    Constraint 13
        speed limit (0.387097) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](4540.22) (constraint not satisfied)
    Constraint 14
        speed limit (0.419355) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](5378.93) (constraint not satisfied)
    Constraint 15
        speed limit (0.451613) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](5977.53) (constraint not satisfied)
    Constraint 16
        speed limit (0.483871) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](6288.68) (constraint not satisfied)
    Constraint 17
        speed limit (0.516129) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](6288.68) (constraint not satisfied)
    Constraint 18
        speed limit (0.548387) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](5977.53) (constraint not satisfied)
    Constraint 19
        speed limit (0.580645) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](5378.93) (constraint not satisfied)
    Constraint 20
        speed limit (0.612903) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](4540.22) (constraint not satisfied)
    Constraint 21
        speed limit (0.645161) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](3532.46)
    Constraint 22
        speed limit (0.677419) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](2466.78)
    Constraint 23
        speed limit (0.709677) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](1618.46)
    Constraint 24
        speed limit (0.741935) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](1010.39)
    Constraint 25
        speed limit (0.774194) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](592.274)
    Constraint 26
        speed limit (0.806452) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](319.695)
    Constraint 27
        speed limit (0.83871) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](154.174)
    Constraint 28
        speed limit (0.870968) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](63.1496)
    Constraint 29
        speed limit (0.903226) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](19.9809)
    Constraint 30
        speed limit (0.935484) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](3.94685)
    Constraint 31
        speed limit (0.967742) (differentiable function)
        Bounds: (-inf, 3612.5)
        Scales: 1
        Initial value: [1](0.246678)
    Starting point: [7](1,0,0,0,200,200,200)
    Starting value: [1](-1)
    Infinity value (for all functions): inf
  Parameters:
    ipopt.acceptable_compl_inf_tol ("acceptance" threshold for the complementarity conditions): 0.01
    ipopt.acceptable_constr_viol_tol ("acceptance" threshold for the constraint violation): 0.01
    ipopt.acceptable_dual_inf_tol ("acceptance" threshold for the dual infeasibility): 1e+10
    ipopt.acceptable_iter (number of "acceptable" iterates before triggering termination): 15
    ipopt.acceptable_tol ("acceptable" convergence tolerance (relative)): 0.05
    ipopt.compl_inf_tol (desired threshold for the complementarity conditions): 0.0001
    ipopt.constr_viol_tol (desired threshold for the constraint violation): 0.0001
    ipopt.derivative_test (enable derivative checker): none
    ipopt.dual_inf_tol (desired threshold for the dual infeasibility): 1
    ipopt.expect_infeasible_problem (enable heuristics to quickly detect an infeasible problem): no
    ipopt.file_print_level (verbosity level for output file): 5
    ipopt.hessian_approximation (): limited-memory
    ipopt.linear_solver (linear solver): mumps
    ipopt.mu_strategy (update strategy for barrier parameter): adaptive
    ipopt.nlp_scaling_max_gradient (maximum gradient after NLP scaling): 100
    ipopt.nlp_scaling_method (technique used for scaling the problem internally before it is solved): gradient-based
    ipopt.option_file_name (file name of options file (to overwrite default)): 
    ipopt.output_file (file name of desired output file (leave unset for no file output)): 
    ipopt.print_level (output verbosity level): 5
    ipopt.print_options_documentation (switch to print all algorithmic options): no
    ipopt.print_user_options (print all options set by the user): no
    ipopt.start_with_resto (tells algorithm to switch to restoration phase in first iteration): no
    ipopt.tol (desired convergence tolerance (relative)): 0.001
    max-iterations (number of iterations): 3000
